pydantic_settings/main.py

```diff
@@ -94,6 +113,7 @@ def settings_customise_sources(
         cls,
         settings_cls: type[BaseSettings],
         init_settings: PydanticBaseSettingsSource,
+        cli_settings: PydanticBaseSettingsSource,
```

Comment:
Please remember we can't change the signature of this function because it will break user code.

Probably, we can add a special `model_validate_cli` that temporarily adds the CLI source at the start of sources, then does validation

Reply:
Yep, this is one of main points of discussion that needed to be had. There are actually a few that need feedback, I'll try and comment the others when I get time.

These are the decisions that needed consideration:
- Do we want break backwards compatibility by adding CLI source as a core source? - No.
  - Benefit here is that CLI source fits into exact same paradigm as other core sources "out of the box".
- Should CLI source have prioritization levels?
  - If yes, it means we will need to have a separate `cli_settings_customise_sources` hook for users to arrange prioritization of CLI sources relative to others
  - Alternatively, we could do something like I outlined [here](https://github.com/pydantic/pydantic-settings/issues/209#issuecomment-1900855951), but it makes it awkward with how args are passed in. Specifically, users have the potential to define args in two different places which I don't like (`_cli_parse_args` and instatiation of `CliSettingsSource`). So, we would need to do something there to make it one location.
- Should the CLI source have a fixed prioritization level, and if so, should it always be the topmost priority?
  - After playing with the implementation, this is the one makes the most sense to me, which is what I think you're advocating for above. Typically users will probably treat `_cli_parse_args` as an alternative to "init_kwargs", or with the intention of topmost priority. This does eliminate use cases where this is not true, but I imagine that is a very small minority.

Reply:
> Should CLI source have prioritization levels?

I would say we can add it as first priority if users want to benefit from it. I've added [an issue](https://github.com/pydantic/pydantic-settings/issues/216) to address the `settings_customise_sources` signature problem in V3

> Should the CLI source have a fixed prioritization level, and if so, should it always be the topmost priority?

Yes, if the user wants to use the CLI source it has to be the top priority source.

Reply:
Sounds good. I'll head in that direction.